File: frontend/app/mocks/reminders.ts
/**
 * Données simulées des RAPPELS — `engagement.reminder_rules` et
 * `engagement.scheduled_reminders` (`110_engagement.sql` § 6).
 *
 * LES QUATRE RAPPELS SONT CUMULÉS, ce n'est pas un choix parmi quatre. Le
 * commanditaire l'a dit ainsi et le modèle l'a écrit ainsi : la colonne
 * `offsets` a pour défaut `{2 days, 1 day, 1 hour, 30 minutes}` et les quatre
 * partent. Un écran qui laisserait choisir « le rappel » au singulier
 * contredirait la règle avant même que l'API n'existe.
 *
 * LES LIGNES NE SONT PAS ÉCRITES À LA MAIN, ELLES SONT DÉRIVÉES — et c'est
 * délibéré. La matérialisation compte une ligne PAR DESTINATAIRE, PAR CANAL ET
 * PAR DÉCALAGE : les sept inscrits d'une seule séance en produisent vingt-huit,
 * et le jeu complet plusieurs centaines. Les recopier serait illisible et faux
 * au premier changement d'inscription. `sessionReminders()` rejoue donc
 * `engagement.schedule_session_reminders()` en TypeScript, exactement comme
 * `organization-search.ts` rejoue `find_similar_organizations()` :
 *
 *   1. règle applicable — celle de la séance si elle existe, sinon celle de
 *      l'édition, SANS CUMUL (`ORDER BY (session_id IS NOT NULL) DESC LIMIT 1`) ;
 *   2. destinataires — inscriptions dont le statut n'est ni `cancelled`, ni
 *      `declined`, ni `waitlisted` ; une personne en liste d'attente n'a pas de
 *      place, la prévenir deux jours avant serait lui en promettre une ;
 *   3. un rappel par (personne × canal × décalage), dont l'instant vaut
 *      `starts_at − offset` ;
 *   4. les rappels dont l'heure est DÉJÀ PASSÉE au moment de la programmation ne
 *      sont pas créés : « on ne réveille personne à 3 h du matin parce qu'un
 *      import a pris du retard », dit le SQL.
 *
 * L'ÉTAT SE DÉDUIT DE L'INSTANT, ce que fait le worker en réalité : parti si
 * l'échéance est derrière nous, en attente sinon. C'est la seule liberté prise
 * avec la base — elle porte un `status` en colonne, écrit par le worker — et
 * elle évite de figer dans le jeu de données une date d'envoi qui vieillirait
 * mal.
 */

import type {
  NotificationChannel,
  OffsetMinutes,
  ReminderRule,
  ScheduledReminder,
} from '~/types/engagement'
import type { ReminderSlot } from '~/types/organization-workspace'
import { EVENT, PERSON, REMINDER_RULE, SCHEDULED_REMINDER } from './ids'
import { registrations } from './registrations'
import { allSessions } from './sessions'

/** Les quatre décalages du modèle, en minutes. */
const DEFAULT_OFFSETS: OffsetMinutes[] = [2 * 24 * 60, 24 * 60, 60, 30]

export const reminderRules = [
  {
    id: REMINDER_RULE.cop31,
    event_id: EVENT.cop31,
    session_id: null,
    offsets: DEFAULT_OFFSETS,
    channels: ['email'],
    type_code: 'programme.session.reminder',
    template_id: null,
    is_active: true,
    created_by: PERSON.perretAdmin,
    created_at: '2026-06-02T09:00:00Z',
    updated_at: '2026-06-02T09:00:00Z',
  },
  {
    id: REMINDER_RULE.cop30,
    event_id: EVENT.cop30,
    session_id: null,
    offsets: DEFAULT_OFFSETS,
    channels: ['email'],
    type_code: 'programme.session.reminder',
    template_id: null,
    is_active: true,
    created_by: PERSON.perretAdmin,
    created_at: '2025-06-04T09:00:00Z',
    updated_at: '2025-06-04T09:00:00Z',
  },
] satisfies ReminderRule[]

/** Statuts d'inscription qui ne reçoivent aucun rappel — cf. la fonction SQL. */
const EXCLUDED_STATUSES = new Set(['cancelled', 'declined', 'waitlisted'])

/**
 * La règle applicable à une séance : celle de la séance d'abord, celle de son
 * édition ensuite. Pas de cumul, pour que l'administrateur sache ce qui part.
 */
function ruleFor(sessionId: string, eventId: string): ReminderRule | null {
  const rules: ReminderRule[] = reminderRules
  return (
    rules.find((rule) => rule.is_active && rule.session_id === sessionId) ??
    rules.find((rule) => rule.is_active && rule.event_id === eventId) ??
    null
  )
}

/**
 * Matérialisation des rappels d'une séance — `schedule_session_reminders()`.
 *
 * Rendue publique parce que le planificateur (A9) et le back-office en auront
 * besoin ligne à ligne. L'espace organisation, lui, consomme l'agrégat
 * ci-dessous : il n'a pas à savoir QUI reçoit quoi.
 */
export function sessionReminders(sessionId: string, at: number = Date.now()): ScheduledReminder[] {
  const sessionIndex = allSessions.findIndex((s) => s.id === sessionId)
  const session = allSessions[sessionIndex]
  if (!session) return []

  const rule = ruleFor(session.id, session.event_id)
  if (!rule) return []

  const recipients = registrations.filter(
    (registration) =>
      registration.session_id === sessionId && !EXCLUDED_STATUSES.has(registration.status),
  )

  const startsAt = Date.parse(session.starts_at)
  const scheduled: ScheduledReminder[] = []
  let sequence = 0

  for (const registration of recipients) {
    for (const offset of rule.offsets) {
      for (const channel of rule.channels as NotificationChannel[]) {
        const scheduledFor = startsAt - offset * 60_000
        // La séance annulée ne rappelle rien : le worker marque `skipped`, et
        // le SQL prévoit le motif (`session_cancelled`).
        const isCancelled = session.status === 'cancelled'
        sequence += 1
        scheduled.push({
          // Numérotés dans l'espace de LEUR séance : deux séances distinctes ne
          // peuvent pas produire le même identifiant, ce qu'un simple compteur
          // local aurait fait dès qu'un écran en charge deux.
          id: SCHEDULED_REMINDER(sessionIndex * 1000 + sequence),
          rule_id: rule.id,
          session_id: session.id,
          person_id: registration.person_id,
          registration_id: registration.id,
          channel,
          offset_before: offset,
          scheduled_for: new Date(scheduledFor).toISOString(),
          status: isCancelled ? 'skipped' : scheduledFor <= at ? 'sent' : 'pending',
          job_id: null,
          sent_at: !isCancelled && scheduledFor <= at ? new Date(scheduledFor).toISOString() : null,
          skip_reason: isCancelled ? 'session_cancelled' : null,
          created_at: registration.created_at,
        })
      }
    }
  }

  return scheduled
}

/**
 * LE CALENDRIER DES RAPPELS D'UNE SÉANCE, tel que son organisation le lit :
 * une ligne par décalage, avec l'état et le nombre de destinataires.
 *
 * C'est l'agrégation que le modèle ne rend pas — obligation d'API inscrite au
 * prompt B6. Elle porte un NOMBRE et jamais une liste : les inscrits d'une
 * séance ne sont pas les données de l'organisation qui l'anime.
 */
export function sessionReminderSchedule(sessionId: string, at: number = Date.now()): ReminderSlot[] {
  const lines = sessionReminders(sessionId, at)
  const groups = new Map<string, ReminderSlot>()

  for (const line of lines) {
    // Clé sur le couple (décalage, canal) : deux canaux d'un même décalage sont
    // deux envois distincts, et les fondre en cacherait un.
    const key = `${line.offset_before}:${line.channel}`
    const existing = groups.get(key)
    if (existing) {
      existing.recipient_count += 1
      continue
    }
    groups.set(key, {
      offset_before: line.offset_before,
      channel: line.channel,
      scheduled_for: line.scheduled_for,
      status: line.status,
      recipient_count: 1,
      sent_at: line.sent_at,
    })
  }

  // Du plus lointain au plus proche du début — l'ordre dans lequel ils partent.
  return [...groups.values()].sort((a, b) => b.offset_before - a.offset_before)
}

File: backend/crates/modules/engagement/src/repo/reminders.rs
//! Le calendrier des rappels d'une séance, **lu par la fonction du modèle**.
//!
//! # Une fonction, deux lecteurs, une seule agrégation
//!
//! `engagement.session_reminder_schedule()` est appelée ici **et** par la
//! composition de l'espace organisation, côté `programme`. Écrites séparément,
//! les deux agrégations divergeraient au premier ajustement de la consolidation
//! — et la divergence serait silencieuse : un nombre de destinataires faux
//! ressemble à un nombre juste (FR-052).
//!
//! # Une fonction qui rend une TABLE ne porte aucune contrainte de nullité
//!
//! SQLx type donc **toutes** ses colonnes en `Option`, y compris celles que la
//! requête ne peut pas rendre nulles. Les annoter une à une est la leçon de B3 :
//! sans cela, la structure porterait sept `Option` et l'écran devrait deviner
//! lesquelles sont de vraies absences. Seuls `skip_reason` et `sent_at` en
//! sont : le premier ne sort que sur un groupe mort, le second tant que rien
//! n'est parti.

use kernel::error::Result;
use sqlx::postgres::PgConnection;
use sqlx::PgPool;
use uuid::Uuid;

use crate::domain::reminder::ReminderSlot;

/// Une ligne par (décalage, canal), rangée du décalage le plus lointain au plus
/// proche — l'ordre que la fonction impose elle-même (FR-050).
pub async fn calendrier(pool: &PgPool, session_id: Uuid) -> Result<Vec<ReminderSlot>> {
    let lignes = sqlx::query!(
        r#"SELECT offset_minutes  AS "offset_minutes!",
                  channel         AS "channel!",
                  scheduled_for   AS "scheduled_for!",
                  status          AS "status!",
                  recipient_count AS "recipient_count!",
                  skip_reason,
                  sent_at
             FROM engagement.session_reminder_schedule($1)"#,
        session_id
    )
    .fetch_all(pool)
    .await?;

    Ok(lignes
        .into_iter()
        .map(|l| ReminderSlot {
            offset_before: l.offset_minutes,
            channel: l.channel,
            scheduled_for: l.scheduled_for,
            status: l.status,
            recipient_count: l.recipient_count,
            skip_reason: l.skip_reason,
            sent_at: l.sent_at,
        })
        .collect())
}

// -----------------------------------------------------------------------------
// Matérialisation, réactivation, décalage, annulation
// -----------------------------------------------------------------------------

/// **La matérialisation passe par la fonction du modèle, et rien d'autre.**
///
/// Elle insère les rappels manquants, **met un travail par rappel en file** et
/// **émet** `engagement.reminders.scheduled`. Le service ne redouble ni l'un ni
/// l'autre : il produirait deux courriels par rappel, et le doublon ne se
/// verrait qu'en production.
pub async fn materialiser(conn: &mut PgConnection, session_id: Uuid) -> Result<i32> {
    let crees = sqlx::query_scalar!(
        r#"SELECT engagement.schedule_session_reminders($1) AS "crees!""#,
        session_id
    )
    .fetch_one(conn)
    .await?;

    Ok(crees)
}

/// **Réactive les rappels d'une inscription reprise** — le cas que la clé
/// d'unicité rend piégeux.
///
/// `ux_scheduled_reminders_once` porte sur (séance, personne, canal, décalage)
/// **sans condition d'état** : la ligne annulée existe toujours, et
/// `ON CONFLICT DO NOTHING` ne la ressuscite pas. Sans cette remise à l'état
/// d'attente, qui se désiste puis revient ne recevrait **plus jamais rien** —
/// sans erreur, sans trace (R21).
///
/// **`job_id` est conservé, et c'est délibéré.** Annuler un rappel ne
/// décommande pas son travail : celui-ci est toujours en file, et il partira à
/// l'heure dite. La condition « l'instant est encore devant » garantit qu'il
/// n'a pas encore tourné — un travail déjà passé aurait un instant derrière
/// nous. Le lui rendre à `NULL` ferait tenter une seconde mise en file que
/// `ux_jobs_idempotency` absorberait, et la ligne resterait sans travail.
pub async fn reactiver(
    conn: &mut PgConnection,
    session_id: Uuid,
    person_id: Option<Uuid>,
) -> Result<u64> {
    let reprises = sqlx::query!(
        "UPDATE engagement.scheduled_reminders
            SET status = 'pending', skip_reason = NULL
          WHERE session_id = $1
            AND ($2::uuid IS NULL OR person_id = $2)
            AND status = 'cancelled'
            AND scheduled_for > now()",
        session_id,
        person_id
    )
    .execute(conn)
    .await?
    .rows_affected();

    Ok(reprises)
}

/// Les rappels **déplacés** d'un créneau qui bouge, et les travaux qu'il faut
/// déplacer avec eux.
pub struct Deplacement {
    pub deplaces: u64,
    /// Les travaux encore en file qui portent ces rappels. Leur échéance vit
    /// dans `platform.jobs`, que ce module n'écrit pas.
    pub job_ids: Vec<Uuid>,
    /// Les rappels dont l'instant est passé sous le nez : écartés, jamais
    /// envoyés en rattrapage.
    pub depasses: u64,
}

/// **Déplace les instants d'envoi, il ne les recrée pas.**
///
/// Recréer se heurterait à la clé d'unicité, qui ne porte pas l'instant : les
/// lignes existantes resteraient à l'ancienne heure et rien ne le dirait. Le
/// décalage est le même pour tous — quand un créneau bouge, tous les rappels
/// qui en dépendent bougent d'autant.
pub async fn decaler(
    conn: &mut PgConnection,
    session_id: Uuid,
    secondes: f64,
    motif_depasse: &str,
) -> Result<Deplacement> {
    let lignes = sqlx::query!(
        r#"UPDATE engagement.scheduled_reminders
              SET scheduled_for = scheduled_for + make_interval(secs => $2)
            WHERE session_id = $1 AND status IN ('pending', 'queued')
        RETURNING id, job_id"#,
        session_id,
        secondes
    )
    .fetch_all(&mut *conn)
    .await?;

    let job_ids: Vec<Uuid> = lignes.iter().filter_map(|l| l.job_id).collect();

    // Un créneau qui recule peut faire passer un décalage derrière nous. On ne
    // rattrape pas : c'est la règle que la fonction du modèle applique déjà à la
    // création, et l'appliquer ici évite qu'un report d'une heure réveille tout
    // le monde d'un coup.
    let depasses = sqlx::query!(
        "UPDATE engagement.scheduled_reminders
            SET status = 'skipped', skip_reason = $2
          WHERE session_id = $1
            AND status IN ('pending', 'queued')
            AND scheduled_for <= now()",
        session_id,
        motif_depasse
    )
    .execute(&mut *conn)
    .await?
    .rows_affected();

    Ok(Deplacement {
        deplaces: lignes.len() as u64,
        job_ids,
        depasses,
    })
}

/// Annule les rappels **encore à traiter** d'une séance, ou d'une personne sur
/// cette séance, **avec leur motif**.
///
/// Les rappels déjà partis ne sont pas touchés : ils sont partis, et le dire
/// autrement serait faux. Le motif est ce qui distingue une annulation d'un
/// oubli — sans lui, l'organisation lirait « rien n'est parti » sans savoir
/// pourquoi (FR-065).
pub async fn annuler(
    conn: &mut PgConnection,
    session_id: Uuid,
    person_id: Option<Uuid>,
    motif: &str,
) -> Result<u64> {
    let annules = sqlx::query!(
        "UPDATE engagement.scheduled_reminders
            SET status = 'cancelled', skip_reason = $3
          WHERE session_id = $1
            AND ($2::uuid IS NULL OR person_id = $2)
            AND status IN ('pending', 'queued')",
        session_id,
        person_id,
        motif
    )
    .execute(conn)
    .await?
    .rows_affected();

    Ok(annules)
}

// -----------------------------------------------------------------------------
// Ce que le travail d'envoi lit et écrit
// -----------------------------------------------------------------------------

/// Un rappel à traiter, avec ce qu'il faut pour décider s'il part.
#[derive(Debug, Clone)]
pub struct RappelATraiter {
    pub id: Uuid,
    pub session_id: Uuid,
    pub person_id: Uuid,
    pub channel: String,
    pub offset_minutes: i32,
    pub status: String,
    /// Le type de notification de la règle qui l'a produit. Replié sur le
    /// défaut du modèle quand la règle a été supprimée — `rule_id` est
    /// `ON DELETE SET NULL`, et un rappel orphelin doit tout de même savoir
    /// quel avis il porte.
    pub type_code: String,
    pub template_id: Option<Uuid>,
}

pub async fn a_traiter(pool: &PgPool, reminder_id: Uuid) -> Result<Option<RappelATraiter>> {
    let ligne = sqlx::query!(
        r#"SELECT sr.id, sr.session_id, sr.person_id,
                  sr.channel::text AS "channel!",
                  round(extract(epoch FROM sr.offset_before) / 60)::int AS "offset_minutes!",
                  sr.status::text AS "status!",
                  COALESCE(r.type_code, 'programme.session.reminder') AS "type_code!",
                  r.template_id
             FROM engagement.scheduled_reminders sr
             LEFT JOIN engagement.reminder_rules r ON r.id = sr.rule_id
            WHERE sr.id = $1"#,
        reminder_id
    )
    .fetch_optional(pool)
    .await?;

    Ok(ligne.map(|l| RappelATraiter {
        id: l.id,
        session_id: l.session_id,
        person_id: l.person_id,
        channel: l.channel,
        offset_minutes: l.offset_minutes,
        status: l.status,
        type_code: l.type_code,
        template_id: l.template_id,
    }))
}

/// **Marque le rappel parti AVANT l'envoi, et rend faux si un autre l'a déjà
/// fait.**
///
/// L'ordre est le cœur de « une fois, et une seule ». La file est « au moins une
/// fois » : un worker tué entre l'envoi et son marquage rejoue le travail
/// entier. Marquer après l'envoi produirait alors **deux courriels** — le défaut
/// exact que la v1 avait, et que la clé d'unicité du modèle ne rattrape pas,
/// puisqu'elle interdit deux lignes, pas deux envois sur la même ligne.
///
/// Le prix est écrit : un worker tué **entre** ce marquage et l'envoi perd un
/// courriel. Un rappel manquant se voit et se rejoue à la main ; un rappel
/// envoyé deux fois est irrattrapable. [`rendre_a_la_file`] limite la perte aux
/// morts brutales, les échecs annoncés étant rendus.
pub async fn marquer_parti(conn: &mut PgConnection, reminder_id: Uuid) -> Result<bool> {
    let pris = sqlx::query!(
        "UPDATE engagement.scheduled_reminders
            SET status = 'sent', sent_at = now(), skip_reason = NULL
          WHERE id = $1 AND status IN ('pending', 'queued')",
        reminder_id
    )
    .execute(conn)
    .await?
    .rows_affected();

    Ok(pris == 1)
}

/// Rend le rappel à la file après un échec d'expédition **annoncé**. Le travail
/// se reprend ; sans ce retour, il trouverait la ligne déjà partie et n'enverrait
/// jamais rien.
pub async fn rendre_a_la_file(conn: &mut PgConnection, reminder_id: Uuid) -> Result<()> {
    sqlx::query!(
        "UPDATE engagement.scheduled_reminders
            SET status = 'queued', sent_at = NULL
          WHERE id = $1 AND status = 'sent'",
        reminder_id
    )
    .execute(conn)
    .await?;
    Ok(())
}

/// **Écarte un rappel avec son motif, jamais en silence** (FR-065).
///
/// Une adresse supprimée, un canal coupé, un canal sans expédition : trois
/// raisons différentes de ne rien envoyer, et l'organisation qui lit son
/// calendrier a droit à laquelle.
pub async fn marquer_ecarte(conn: &mut PgConnection, reminder_id: Uuid, motif: &str) -> Result<()> {
    sqlx::query!(
        "UPDATE engagement.scheduled_reminders
            SET status = 'skipped', skip_reason = $2
          WHERE id = $1 AND status IN ('pending', 'queued')",
        reminder_id,
        motif
    )
    .execute(conn)
    .await?;
    Ok(())
}
